language/move-prover/docgen/src/docgen.rs

```diff
@@ -664,7 +688,76 @@ impl<'env> Docgen<'env> {
         }
     }
 
-    /// Gen header for TOC, returning label where we can later insert the content after
+    /// Generate a .svg file of the forward (or backward) dependency diagram for the given module.
+    fn gen_dependency_diagram(&self, module_id: ModuleId, is_forward: bool) {
+        let module_env = self.env.get_module(module_id);
+        let module_name = module_env.get_name().display(module_env.symbol_pool());
+
+        let mut dot_src: String = String::from("digraph G {{\n");
+        let mut visited: BTreeSet<ModuleId> = BTreeSet::new();
+        let mut queue: VecDeque<ModuleId> = VecDeque::new();
+
+        visited.insert(module_id);
+        queue.push_back(module_id);
+
+        while let Some(id) = queue.pop_front() {
+            let env = self.env.get_module(id);
+            let name = env.get_name().display(env.symbol_pool());
+            let dep_list = if is_forward {
+                env.get_used_modules(false)
+            } else {
+                env.get_using_modules(false)
+            };
+            dot_src.push_str(&format!("    {}\n", name));
+            for dep_id in dep_list.iter() {
+                if id == *dep_id {
+                    continue;
+                }
+                let dep_env = self.env.get_module(*dep_id);
+                let dep_name = dep_env.get_name().display(dep_env.symbol_pool());
+                if is_forward {
+                    dot_src.push_str(&format!("    {} -> {}\n", name, dep_name));
+                } else {
+                    dot_src.push_str(&format!("    {} -> {}\n", dep_name, name));
+                }
+                if !visited.contains(dep_id) {
+                    visited.insert(*dep_id);
+                    queue.push_back(*dep_id);
+                }
+            }
+        }
+        dot_src.push_str("}}\n");
+
+        let mut out_path = PathBuf::from(&self.options.output_directory).join("img");
+        fs::create_dir_all(&out_path).unwrap(); // TODO: error handling
```

Comment:
In this function, there are a few places where error handling needs to be done properly. What would be a good way to handle the errors in this case? `gen_dependency_diagram` and its callers (e.g., `gen_modules(...)` and `gen()`) don't return the Result type (e.g., anyhow::Result<()>).

Reply:
In general, we add errors to the global env, and then either continue and compensate, or gracefully terminate (preferable the first one). There are examples, e.g. search for the string ""cannot read root template" to see one. The overall driver will print all the errors at the end of the run.